src/handlers/netprobe/NetProbeStreamHandler.cpp

```diff
@@ -0,0 +1,208 @@
+    }
+
+    if (config_exists("recorded_stream")) {
+        _metrics->set_recorded_stream();
+    }
+
+    if (_netprobe_proxy) {
+        _probe_send_connection = _netprobe_proxy->probe_send_signal.connect(&NetProbeStreamHandler::probe_signal_send, this);
+        _probe_recv_connection = _netprobe_proxy->probe_recv_signal.connect(&NetProbeStreamHandler::probe_signal_recv, this);
+        _probe_fail_connection = _netprobe_proxy->probe_fail_signal.connect(&NetProbeStreamHandler::probe_signal_fail, this);
+        _heartbeat_connection = _netprobe_proxy->heartbeat_signal.connect(&NetProbeStreamHandler::check_period_shift, this);
+    }
+
+    _running = true;
+}
+
+void NetProbeStreamHandler::stop()
+{
+    if (!_running) {
+        return;
+    }
+
+    if (_netprobe_proxy) {
+        _probe_send_connection.disconnect();
+        _probe_recv_connection.disconnect();
+        _probe_fail_connection.disconnect();
+        _heartbeat_connection.disconnect();
+    }
+
+    _running = false;
+}
+
+void NetProbeStreamHandler::probe_signal_send(pcpp::Packet &payload, TestType type, const std::string &name, timespec stamp)
+{
+    if (type == TestType::Ping) {
+        if (auto icmp = payload.getLayerOfType<pcpp::IcmpLayer>(); icmp != nullptr) {
+            _metrics->process_netprobe_icmp(icmp, name, stamp);
+        }
+    }
+}
+
+void NetProbeStreamHandler::probe_signal_recv(pcpp::Packet &payload, TestType type, const std::string &name, timespec stamp)
+{
+    if (type == TestType::Ping) {
+        if (auto icmp = payload.getLayerOfType<pcpp::IcmpLayer>(); icmp != nullptr) {
+            _metrics->process_netprobe_icmp(icmp, name, stamp);
+        }
+    }
+}
+
+void NetProbeStreamHandler::probe_signal_fail([[maybe_unused]] ErrorType error, [[maybe_unused]] TestType type, [[maybe_unused]] const std::string &name)
+{
+}
+
+void NetProbeMetricsBucket::specialized_merge(const AbstractMetricsBucket &o, Metric::Aggregate agg_operator)
+{
+    // static because caller guarantees only our own bucket type
+    const auto &other = static_cast<const NetProbeMetricsBucket &>(o);
+
+    std::shared_lock r_lock(other._mutex);
+    std::unique_lock w_lock(_mutex);
+
+    for (const auto &target : other._targets_metrics) {
+        const auto &targetId = target.first;
+        if (!_targets_metrics.count(targetId)) {
+            _targets_metrics[targetId] = std::make_unique<Target>();
+        }
+
+        _targets_metrics[targetId]->attempts += target.second->attempts;
+        _targets_metrics[targetId]->successes += target.second->successes;
+        _targets_metrics[targetId]->time_us.merge(target.second->time_us, agg_operator);
+    }
+}
+
+void NetProbeMetricsBucket::to_prometheus(std::stringstream &out, Metric::LabelMap add_labels) const
+{
+    std::shared_lock r_lock(_mutex);
+
+    for (const auto &target : _targets_metrics) {
+        auto target_labels = add_labels;
+        auto targetId = target.first;
+        target_labels["target"] = targetId;
```

Comment:
is targetId always an IP, or will it be a hostname if they specified it that way in the policy?

Reply:
actually it will be the specified name. specified on the policy
```
....
          targets:
            test_1_name:
              target: foo.bar
            test_2_name:
              target: 10.0.0.1
              tos: EF
```


should we have two labels? one for target name and other for the target? If so... how should we show that on `json` return

Reply:
ah interesting. ok i think it's ok to land it this way - this interface can be considered beta for now anyway. but it provides a way to keep the label consistent in the dashboards, which is what I was concerned about 👍